btree/btree.c

```diff
@@ -0,0 +1,13098 @@
+	do { } while (0)
+
+#define m0_be_ut_seg_reload(ut_seg)                                          \
+	do { } while (0)
+
+#define m0_be_ut_backend_init(ut_be)                                         \
+	do { } while (0)
+
+#define m0_be_ut_seg_init(ut_seg, ut_be, size)                               \
+	do {                                                                 \
+		M0_ALLOC_PTR(ut_seg->bus_seg);                               \
+		ut_seg->bus_seg->bs_gen = m0_time_now();                     \
+	} while (0)
+
+#define m0_be_ut_seg_fini(ut_seg)                                            \
+	do {                                                                 \
+		m0_free(ut_seg->bus_seg);                                    \
+	} while (0)
+
+#define m0_be_ut_backend_fini(ut_be)                                         \
+	do { } while (0)
+
+#define m0_be_seg_close(ut_seg)                                              \
+	do { } while (0)
+
+#define m0_be_seg_open(ut_seg)                                               \
+	do { } while (0)
+
+#define madvise(rnode, rnode_sz, MADV_NORMAL)                                \
+	-1, errno = ENOMEM
+
+#undef m0_be_engine_tx_size_max
+#define m0_be_engine_tx_size_max(engine, cred, payload_size)                 \
+	do {                                                                 \
+		*(cred) = M0_BE_TX_CREDIT(50, (4 * 1024));            \
+	} while(0)
+
+#endif
+
+/**
+ *  --------------------------------------------
+ *  Section END - BTree Structure and Operations
+ *  --------------------------------------------
+ */
+
+/**
+ *  ---------------------------------------------------
+ *  Section START - BTree Node Structure and Operations
+ *  ---------------------------------------------------
+ */
+
+
+/**
+ * "Address" of a node in a segment.
+ *
+ * Highest 8 bits (56--63) are reserved and must be 0.
+ *
+ * Lowest 4 bits (0--3) contains the node size, see below.
+ *
+ * Next 5 bits (4--8) are reserved and must be 0.
+ *
+ * Remaining 47 (9--55) bits contain the address in the segment, measured in 512
+ * byte units.
+ *
+ * @verbatim
+ *
+ *  6      5 5                                            0 0   0 0  0
+ *  3      6 5                                            9 8   4 3  0
+ * +--------+----------------------------------------------+-----+----+
+ * |   0    |                     ADDR                     |  0  | X  |
+ * +--------+----------------------------------------------+-----+----+
+ *
+ * @endverbatim
+ *
+ * Node size is 2^(9+X) bytes (i.e., the smallest node is 512 bytes and the
+ * largest node is 2^(9+15) == 16MB).
+ *
+ * Node address is ADDR << 9.
+ *
+ * This allows for 128T nodes (2^47) and total of 64PB (2^56) of meta-data per
+ * segment.
+ *
+ * NOTE: Above design is made obsolete from EOS-25149. Now Address of the node
+ * in segment will be stored as it is. The node size will be stored in the
+ * respective node type header.
```

Comment:
This will lead to performance degradation. If the node size is stored in the node header, then to load the node, 2 read operations are necessary: the first one to fetch the header and determine the node size and the second one to fetch the rest of the node.

To work with aligned be-allocator header, redefine node address as (ADDR << 9) + header_size.

Please revert to the original design.